### PR TITLE
feat(tjs): stream JSON scripts to import ER data

### PR DESCRIPTION
**Pull request by jbranchaud**
Add scripts to import both Purchase data (including users, coupons, etc.) and Lesson Progress data for Epic React users.

These scripts work with the JSON data produced by the scripts added in https://github.com/skillrecordings/egghead-rails/pull/5176.

Because the JSON files are so large, they cannot be read in as a single string of data (`ERR_STRING_TOO_LONG`), so I updated these scripts (originally based on the TJS ones) to stream the JSON.

These will be run as part of the hard cutover for Epic React detailed here: https://roamresearch.com/#/app/egghead/page/SlsmZU15S

Details on the timing of these scripts can be found in Roam: https://roamresearch.com/#/app/egghead/page/mC7MHA8FM

stream

### CODE REVIEW

File: apps/epic-react/src/data/lesson-progress-import.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      await prisma.lessonProgress.createMany({
+        data: validLessonProgressRecords,
+        skipDuplicates: true,
+      })
+    }
+  }
+
+  const collectData = async (data: any) => {
+    await processData(data.value)
+  }
```
Comment (jbranchaud) [inline apps/epic-react/src/data/lesson-progress-import.ts:108]: as the JSON streams in, this function gets called which sends the data to get processed.

File: apps/epic-react/src/data/lesson-progress-import.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  // ***************************************
+  // ** Read and Parse Full Purchase JSON **
+  // ***************************************
+
+  const processData = async (dataToParse: any) => {
+    let {lessonProgress} = FileDataSchema.parse(dataToParse)
+
+    const shortRun = false
+
+    const chunkData = <T>(data: Array<T>): Array<T>[] => {
```
Comment (jbranchaud) [inline apps/epic-react/src/data/lesson-progress-import.ts:82]: process the streamed JSON -- basically parse and chunk the data and then bulk insert with Prisma into the database

File: apps/epic-react/src/data/lesson-progress-import.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  const pipeline = chain([
+    fs.createReadStream(dataFilePath),
+    parser(),
+    streamValues(),
+  ])
+
+  // Aggregate all the data and then process it (push to database)
+  pipeline.on('data', collectData)
+  pipeline.on('end', () => console.log('Done streaming this JSON file'))
+  // ---------------------------------------
```
Comment (jbranchaud) [inline apps/epic-react/src/data/lesson-progress-import.ts:121]: this is the bit of `stream-json` library setup that steams the file, parses it json, and processes it in a streaming pipeline

File: apps/epic-react/src/data/purchase-import.ts [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+  const pipeline = chain([
+    fs.createReadStream(dataFilePath),
+    parser(),
+    streamValues(),
+  ])
+
+  // Aggregate all the data and then process it (push to database)
+  pipeline.on('data', collectData)
+  pipeline.on('end', processData)
+  // ---------------------------------------
```
Comment (jbranchaud) [inline apps/epic-react/src/data/purchase-import.ts:294]: pretty much same story in this file regarding the data be streamed in as JSON, see comments on the Lesson Progress import script for more details.

**Review by joelhooks — APPROVED** [review]
looks great, the streaming approach is nice

**Review by vojtaholik — APPROVED** [review]

**Pull request merged**